packages/std/src/coins.rs

```diff
@@ -0,0 +1,380 @@
+use std::collections::BTreeMap;
+use std::fmt;
+use std::str::FromStr;
+
+use crate::{errors::CoinsError, Coin, StdError, StdResult, Uint128};
+use crate::{OverflowError, OverflowOperation};
+
+/// A collection of coins, similar to Cosmos SDK's `sdk.Coins` struct.
+///
+/// Differently from `sdk.Coins`, which is a vector of `sdk.Coin`, here we
+/// implement Coins as a BTreeMap that maps from coin denoms to amounts.
+/// This has a number of advantages:
+///
+/// - coins are naturally sorted alphabetically by denom
+/// - duplicate denoms are automatically removed
+/// - cheaper for searching/inserting/deleting: O(log(n)) compared to O(n)
+#[derive(Clone, Default, Debug, PartialEq, Eq)]
+pub struct Coins(BTreeMap<String, Uint128>);
+
+/// Casting a Vec<Coin> to Coins.
+/// The Vec can be out of order, but must not contain duplicate denoms.
+/// If you want to sum up duplicates, create an empty instance using `Coins::default` and
+/// use `Coins::add` to add your coins.
+impl TryFrom<Vec<Coin>> for Coins {
+    type Error = CoinsError;
+
+    fn try_from(vec: Vec<Coin>) -> Result<Self, CoinsError> {
+        let mut map = BTreeMap::new();
+        for Coin { amount, denom } in vec {
+            if amount.is_zero() {
+                continue;
+            }
+
+            // if the insertion returns a previous value, we have a duplicate denom
+            if map.insert(denom, amount).is_some() {
+                return Err(CoinsError::DuplicateDenom);
+            }
+        }
+
+        Ok(Self(map))
+    }
+}
+
+impl TryFrom<&[Coin]> for Coins {
+    type Error = CoinsError;
+
+    fn try_from(slice: &[Coin]) -> Result<Self, CoinsError> {
+        slice.to_vec().try_into()
+    }
+}
+
+impl From<Coin> for Coins {
+    fn from(value: Coin) -> Self {
+        let mut coins = Coins::default();
+        // this can never overflow (because there are no coins in there yet), so we can unwrap
+        coins.add(value).unwrap();
+        coins
```

Comment:
Implementation looks good. Could you add a few tests for that (especially happy path and 0 coin leads to an empty Coins)